File: src/main/java/me/ctknight/uploadmanager/UploadNotifier.java
/*
 * Copyright (c) 2016. All rights reserved. Lai Jiewen <alanljw12345@gmail.com>
 */

package me.ctknight.uploadmanager;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.net.Uri;
import android.os.Build;
import android.os.SystemClock;
import android.support.v4.util.LongSparseArray;
import android.support.v4.app.NotificationCompat;
import android.text.TextUtils;
import android.util.Log;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import me.ctknight.uploadmanager.util.TimeUtils;

public class UploadNotifier {

    public static final String NOTIFICATION_CHANNEL = "Upload notification";

    private static final int TYPE_ACTIVE = 1;
    private static final int TYPE_WAITING = 2;
    private static final int TYPE_COMPLETE = 3;
    private static final int TYPE_FAILED = 4;


    private static Context mContext = null;
    //AOSP use final ,but I have to get current package name in a static method,so change it to static.
    private final NotificationManager mNotifManager;

    private final HashMap<String, Long> mActiveNotif = new HashMap<>();

    private final LongSparseArray<Long> mUploadSpeed = new LongSparseArray<>();

    private final LongSparseArray<Long> mUploadTouch = new LongSparseArray<>();
    //In AOSP Download Notifier,they use LongSparseLongArray,
    //actually,LongSparseArray is a generic version of LongSparseLongArray,
    //so LongSparseArray<Long> is totally same with LongSparseLongArray.


    public UploadNotifier(Context context) {
        mContext = context;
        mNotifManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && mNotifManager != null) {
            NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL,
                    mContext.getString(R.string.notification_channel),
                    NotificationManager.IMPORTANCE_DEFAULT);
            mNotifManager.createNotificationChannel(channel);
        }
    }

    private static CharSequence getUploadTitle(Resources res, UploadInfo info) {
        if (!TextUtils.isEmpty(info.mTitle)) {
            return info.mTitle;
        } else {
            return res.getString(R.string.upload_unknown_upload_title);
        }
    }

    private static String buildNotificationTag(UploadInfo info) {
        // waiting and active is clustered
        if (UploadContract.isWaiting(info)) {
            return TYPE_WAITING + ":" + mContext.getPackageName();
        } else if (UploadContract.isOnGoing(info)) {
            return TYPE_ACTIVE + ":" + mContext.getPackageName();
        } else if (UploadContract.isFailed(info)) {
            return TYPE_FAILED + ":" + info.mId;
        } else if (UploadContract.isComplete(info)) {
            return TYPE_COMPLETE + ":" + info.mId;
        } else {
            return null;
        }
    }

    private static int getNotificationTagType(String tag) {
        return Integer.parseInt(tag.substring(0, tag.indexOf(':')));
    }

    public void cancelAll() {
        mNotifManager.cancelAll();
    }

    public void notifyUploadSpeed(long id, long bytesPerSecond) {
        synchronized (mUploadSpeed) {
            if (bytesPerSecond != 0) {
                mUploadSpeed.put(id, bytesPerSecond);
                mUploadTouch.put(id, SystemClock.elapsedRealtime());
            } else {
                mUploadSpeed.delete(id);
                mUploadTouch.delete(id);
            }
        }
    }

    public void updateWith(Collection<UploadInfo> uploads) {
        synchronized (mActiveNotif) {
            updateWithLocked(uploads);
        }
    }

    public void updateWithLocked(Collection<UploadInfo> uploads) {
        final Resources res = mContext.getResources();

        // Cluster uploads together
        final Map<String, ArrayList<UploadInfo>> clustered = new HashMap<>();
//        final Multimap<String, UploadInfo> clustered = ArrayListMultimap.create();
        for (UploadInfo info : uploads) {
            final String tag = buildNotificationTag(info);
            if (tag != null) {
                if (clustered.containsKey(tag)) {
                    clustered.get(tag).add(info);
                } else {
                    //NOTE I know this is dirty, but I just don't want to get guava involved.
                    clustered.put(tag, new ArrayList<>(Arrays.asList(info)));
                }
            }
        }

        //Build notification for each cluster
        for (String tag : clustered.keySet()) {
            final int type = getNotificationTagType(tag);
            final Collection<UploadInfo> cluster = clustered.get(tag);

            final NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, NOTIFICATION_CHANNEL);
            builder.setColor(res.getColor(R.color.system_notification_accent_color));

            // Use time when cluster was first shown to avoid shuffling
            final long firstShown;
            if (mActiveNotif.containsKey(tag)) {
                firstShown = mActiveNotif.get(tag);
            } else {
                firstShown = System.currentTimeMillis();
                mActiveNotif.put(tag, firstShown);
            }
            builder.setWhen(firstShown);

            // Show relevant icon
            if (type == TYPE_ACTIVE) {
                builder.setSmallIcon(android.R.drawable.stat_sys_upload);
            } else if (type == TYPE_WAITING) {
                builder.setSmallIcon(android.R.drawable.stat_sys_warning);
            } else if (type == TYPE_COMPLETE) {
                builder.setSmallIcon(android.R.drawable.stat_sys_upload_done);
            } else if (type == TYPE_FAILED) {
                builder.setSmallIcon(android.R.drawable.stat_sys_warning);
            }

            // Build action intents
            // add action by type
            // active -> cancel
            // fail waiting -> retry
            if (type == TYPE_ACTIVE || type == TYPE_WAITING) {
                final Uri uri = new Uri.Builder().scheme("active-dl").appendPath(tag).build();
                final Intent intent = new Intent(UploadContract.ACTION_LIST,
                        uri, mContext, UploadReceiver.class);
                intent.putExtra(UploadManager.EXTRA_NOTIFICATION_CLICK_UPLOAD_IDS,
                        getUploadIds(cluster));
                builder.setContentIntent(PendingIntent.getBroadcast(mContext,
                        9, intent, PendingIntent.FLAG_UPDATE_CURRENT));
                builder.setOngoing(true);

                final UploadInfo info = cluster.iterator().next();
                if (type == TYPE_ACTIVE) {
                    final Uri idUri = ContentUris.withAppendedId(UploadContract.UPLOAD_URIS.CONTENT_URI, info.mId);
                    final Intent actionIntent = new Intent(UploadContract.ACTION_CANCEL,
                            idUri, mContext, UploadReceiver.class);
                    builder.addAction(R.drawable.ic_clear_black_24dp,
                            mContext.getString(R.string.notification_action_cancel),
                            PendingIntent.getBroadcast(mContext, 0, actionIntent, 0));
                } else {
                    // WAITING
                    final Uri idUri = ContentUris.withAppendedId(UploadContract.UPLOAD_URIS.CONTENT_URI, info.mId);
                    final Intent actionIntent = new Intent(UploadContract.ACTION_MANUAL_REDO,
                            idUri, mContext, UploadReceiver.class);
                    builder.addAction(R.drawable.ic_redo_black_24dp,
                            mContext.getString(R.string.notification_action_redo),
                            PendingIntent.getBroadcast(mContext, 0, actionIntent, 0));
                }

            } else if (type == TYPE_COMPLETE) {
                final UploadInfo info = cluster.iterator().next();
                final Uri uri = ContentUris.withAppendedId(UploadContract.UPLOAD_URIS.CONTENT_URI, info.mId);
                builder.setAutoCancel(true);
                //In AOSP, action is set according to its status
                //but we just need to open the upload list.

                final String action;
                if (UploadContract.isStatusError(info.mStatus)) {
                    action = UploadContract.ACTION_LIST;
                } else {
                    action = UploadContract.ACTION_OPEN;
                }
                final Intent intent = new Intent(action, uri, mContext, UploadReceiver.class);
                intent.putExtra(UploadManager.EXTRA_NOTIFICATION_CLICK_UPLOAD_IDS,
                        getUploadIds(cluster));
                builder.setContentIntent(PendingIntent.getBroadcast(mContext,
                        0, intent, PendingIntent.FLAG_UPDATE_CURRENT));

                final Intent hideIntent = new Intent(UploadContract.ACTION_HIDE,
                        uri, mContext, UploadReceiver.class);
                builder.setDeleteIntent(PendingIntent.getBroadcast(mContext, 0, hideIntent, 0));

            } else if (type == TYPE_FAILED) {
                final UploadInfo info = cluster.iterator().next();
                final Uri uri = ContentUris.withAppendedId(UploadContract.UPLOAD_URIS.CONTENT_URI, info.mId);
                builder.setAutoCancel(true);

                final String action = UploadContract.ACTION_LIST;

                final Intent intent = new Intent(action, uri, mContext, UploadReceiver.class);
                intent.putExtra(UploadManager.EXTRA_NOTIFICATION_CLICK_UPLOAD_IDS,
                        getUploadIds(cluster));
                builder.setContentIntent(PendingIntent.getBroadcast(mContext,
                        0, intent, PendingIntent.FLAG_UPDATE_CURRENT));

                final Intent hideIntent = new Intent(UploadContract.ACTION_HIDE,
                        uri, mContext, UploadReceiver.class);
                builder.setDeleteIntent(PendingIntent.getBroadcast(mContext, 0, hideIntent, 0));
            }

            //Calculate and show progress
            String remainingText = null;
            String percentText = null;
            if (type == TYPE_ACTIVE) {
                long current = 0;
                long total = 0;
                long speed = 0;
                synchronized (mUploadSpeed) {
                    for (UploadInfo info : cluster) {
                        if (info.mTotalBytes != -1) {
                            current += info.mCurrentBytes;
                            total += info.mTotalBytes;
                            speed += mUploadSpeed.get(info.mId) == null ? 0 : mUploadSpeed.get(info.mId);
                        }
                    }
                }

                if (total > 0) {
                    percentText =
                            NumberFormat.getPercentInstance().format((double) current / total);

                    if (speed > 0) {
                        final long remainingMillis = ((total - current) * 1000) / speed;
                        remainingText = res.getString(R.string.upload_remaining,
                                TimeUtils.formatDuration(remainingMillis, res));
                    }

                    final int percent = (int) ((current * 100) / total);
                    builder.setProgress(100, percent, false);
                } else {
                    builder.setProgress(100, 0, true);
                }
            }

            final Notification notification;
            if (cluster.size() == 1) {
                final UploadInfo info = cluster.iterator().next();

                builder.setContentTitle(getUploadTitle(res, info));

                if (type == TYPE_ACTIVE) {
                    if (!TextUtils.isEmpty(info.mDescription)) {
                        builder.setContentText(info.mDescription);
                    } else {
                        builder.setContentText(remainingText);
                    }
                    builder.setContentInfo(percentText);

                } else if (type == TYPE_WAITING) {
                    //For both continuable transfer and network check before transfer,
                    //but Box doesn't support previous one.
                    builder.setContentText(res.getText(
                            R.string.notification_waiting_for_suitable_network));

                } else if (type == TYPE_COMPLETE) {
                    builder.setContentText(res.getString(R.string.notification_upload_successfully));

                } else if (type == TYPE_FAILED) {
                    // TODO: 2016/2/28 TEMP FIX
                    if (info.mStatus == UploadContract.UPLOAD_STATUS.FILE_NOT_FOUND) {
                        builder.setContentText(res.getString(R.string.notification_upload_file_not_found));
                    } else {
                        builder.setContentText(res.getString(R.string.notification_upload_unsuccessfully));
                    }
                }

                notification = builder.build();

            } else {
                // handle more than one items

                final NotificationCompat.InboxStyle inboxStyle = new NotificationCompat.InboxStyle(builder);

                for (UploadInfo info : cluster) {
                    inboxStyle.addLine(getUploadTitle(res, info));
                }

                if (type == TYPE_ACTIVE) {
                    builder.setContentTitle(res.getQuantityString(
                            R.plurals.notif_summary_active, cluster.size(), cluster.size()));
                    builder.setContentText(remainingText);
                    builder.setContentInfo(percentText);
                    inboxStyle.setSummaryText(remainingText);

                } else if (type == TYPE_WAITING) {
                    builder.setContentTitle(
                            res.getQuantityString(R.plurals.notif_summary_waiting, cluster.size(), cluster.size()));
                    builder.setContentText(
                            res.getString(R.string.notification_waiting_for_suitable_network));
                    inboxStyle.setSummaryText(
                            res.getString(R.string.notification_waiting_for_suitable_network));

                } else if (type == TYPE_FAILED) {
                    builder.setContentText(res.getQuantityString(R.plurals.notif_summary_failed, cluster.size(), cluster.size()));
                }

                notification = inboxStyle.build();

            }
            mNotifManager.notify(tag, 0, notification);

        }

        final Iterator<String> it = mActiveNotif.keySet().iterator();
        while (it.hasNext()) {
            final String tag = it.next();
            if (!clustered.containsKey(tag)) {
                mNotifManager.cancel(tag, 0);
                it.remove();
            }
        }
    }

    private long[] getUploadIds(Collection<UploadInfo> infos) {
        final long[] ids = new long[infos.size()];
        int i = 0;
        for (UploadInfo info : infos) {
            ids[i++] = info.mId;
        }
        return ids;
    }

    public void dumpSpeeds() {
        synchronized (mUploadSpeed) {
            for (int i = 0; i < mUploadSpeed.size(); i++) {
                final long id = mUploadSpeed.keyAt(i);
                final long delta = SystemClock.elapsedRealtime() - mUploadTouch.get(id);
                Log.d("UploadManager", "Upload " + id + " speed " + mUploadSpeed.valueAt(i) + "bps, "
                        + delta + "ms ago");

            }
        }
    }
}

File: lib/src/main/java/me/ctknight/uploadmanager/internal/UploadNotifier.kt
/*
 * Copyright (c) 2019. All rights reserved. Lai Jiewen <alanljw12345@gmail.com>
 */

package me.ctknight.uploadmanager.internal

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.net.Uri
import android.os.Build
import android.os.SystemClock
import android.util.Log
import androidx.annotation.GuardedBy
import androidx.collection.ArrayMap
import androidx.collection.LongSparseArray
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.getSystemService
import androidx.core.content.res.ResourcesCompat
import me.ctknight.uploadmanager.*
import me.ctknight.uploadmanager.thirdparty.SingletonHolder
import me.ctknight.uploadmanager.util.LogUtils
import me.ctknight.uploadmanager.util.formatDuration
import java.text.NumberFormat


//In AOSP Download Notifier,they use LongSparseLongArray,
//actually,LongSparseArray is a generic version of LongSparseLongArray,
//so LongSparseArray<Long> is totally same with LongSparseLongArray.
internal class UploadNotifier(private val mContext: Context) {
  private val resources: Resources
    get() = mContext.resources
  //AOSP use final ,but I have to get current package name in a static method,so change it to static.
  private val mNotifManager: NotificationManagerCompat = NotificationManagerCompat.from(mContext)

  private val mDatabase: UploadDatabase = Database.getInstance(mContext)

  // {tag: firstShown}
  @GuardedBy("mActiveNotif")
  private val mActiveNotif: MutableMap<Pair<Long, NotificationStatus>, Long> = ArrayMap()

  // {Id: {Speed, lastModified}}
  @GuardedBy("mUploadSpeed")
  private val mUploadSpeed = LongSparseArray<Pair<Long, Long>>()

  init {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val notificationManager: NotificationManager = mContext.getSystemService()!!
      with(notificationManager) {
        createNotificationChannel(NotificationChannel(CHANNEL_ACTIVE,
            mContext.getText(R.string.upload_running),
            NotificationManager.IMPORTANCE_MIN))
        createNotificationChannel(NotificationChannel(CHANNEL_WAITING,
            mContext.getText(R.string.upload_queued),
            NotificationManager.IMPORTANCE_DEFAULT))
        createNotificationChannel(NotificationChannel(CHANNEL_COMPLETE,
            mContext.getText(R.string.upload_complete), NotificationManager.IMPORTANCE_DEFAULT))
      }
    }
  }

  fun notifyUploadSpeed(id: Long, bytesPerSecond: Long) {
    synchronized(mUploadSpeed) {
      if (bytesPerSecond != 0L) {
        mUploadSpeed.put(id, Pair(bytesPerSecond, SystemClock.elapsedRealtime()))
      } else {
        mUploadSpeed.delete(id)
      }
    }
  }

  fun update() {
    synchronized(mActiveNotif) {
      val uploads = mDatabase.uploadManagerQueries.selectAll().executeAsList()
      updateWithLocked(uploads)
    }
  }

  private fun updateWithLocked(uploads: Iterable<UploadRecord>) {

    // Cluster uploads together
    // waiting and active is clustered (ongoing)
    // complete and failed is not (can be swiped)

    val clustered = uploads
        .filter {
          it.Status != UploadContract.UploadStatus.DELETED &&
              (it.Visibility == UploadContract.Visibility.VISIBLE ||
                  (it.Visibility == UploadContract.Visibility.HIDDEN_UNTIL_COMPLETE
                      && it.Status.isTerminated()))
        }
        .groupBy {
          val notificationStatus = it.notificationStatus()
          return@groupBy when (notificationStatus) {
            NotificationStatus.WAITING ->
              Pair(-1L, notificationStatus)
            else ->
              Pair(it._ID, notificationStatus)
          }
        }

    clustered.entries.forEach { (tag, cluster) ->
      buildClusterNotification(tag, cluster)
    }
    // remove stale notifications
    val staleKeys = mActiveNotif.keys - clustered.keys
    staleKeys.forEach {
      mNotifManager.cancel(it.toString(), 0)
      mActiveNotif.remove(it)
    }
  }

  private fun buildClusterNotification(tag: Pair<Long, NotificationStatus?>, cluster: List<UploadRecord>) {
    val (id, type) = tag
    type ?: return
    if (cluster.isEmpty()) {
      return
    }
    val res = mContext.resources

    val channel = when (type) {
      NotificationStatus.ACTIVE -> CHANNEL_ACTIVE
      NotificationStatus.COMPLETE -> CHANNEL_COMPLETE
      NotificationStatus.WAITING -> CHANNEL_WAITING
    }
    val builder = NotificationCompat.Builder(mContext, channel)

    // Show relevant icon
    when (type) {
      NotificationStatus.ACTIVE -> builder.setSmallIcon(android.R.drawable.stat_sys_upload)
      NotificationStatus.COMPLETE -> builder.setSmallIcon(android.R.drawable.stat_sys_upload_done)
      NotificationStatus.WAITING -> builder.setSmallIcon(android.R.drawable.stat_sys_warning)
    }

    builder.color = ResourcesCompat.getColor(res, R.color.system_notification_accent_color, null)
    // Use time when cluster was first shown to avoid shuffling
    // TODO: should firstShown be stored in to database
    val firstShown = mActiveNotif.getOrPut(Pair(id, type)) { System.currentTimeMillis() }
    builder.setWhen(firstShown)
    builder.setOnlyAlertOnce(true)

    // Calculate and show progress
    val uploadIds = getUploadIds(cluster)
    if (type == NotificationStatus.ACTIVE || type == NotificationStatus.WAITING) {
      val uri = Uri.Builder().scheme("active-ul")
          .appendPath(UploadContract.UPLOAD_CONTENT_URI.toString()).appendPath(tag.toString()).build()
      val intent = Intent(UploadContract.NotificationAction.List.actionString,
          uri, mContext, UploadReceiver::class.java)
      intent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND)
      intent.putExtra(UploadManager.EXTRA_NOTIFICATION_CLICK_UPLOAD_IDS, uploadIds)
      intent.putExtra(UploadReceiver.EXTRA_CANCELED_UPLOAD_NOTIFICATION_TAG, tag.toString())
      builder.setContentIntent(PendingIntent.getBroadcast(mContext,
          0, intent, PendingIntent.FLAG_UPDATE_CURRENT))

      if (type == NotificationStatus.ACTIVE) {
        builder.setOngoing(true)
      }

      val cancelUri = Uri.Builder().scheme("cancel-ul")
          .appendPath(UploadContract.UPLOAD_CONTENT_URI.toString()).appendPath(tag.toString()).build()
      val cancelIntent = Intent(
          UploadContract.NotificationAction.Cancel.actionString,
          cancelUri,
          mContext,
          UploadReceiver::class.java
      )
      cancelIntent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND)
      cancelIntent.putExtra(UploadReceiver.EXTRA_CANCELED_UPLOAD_IDS, uploadIds.clone())
      cancelIntent.putExtra(UploadReceiver.EXTRA_CANCELED_UPLOAD_NOTIFICATION_TAG, tag.toString())

      builder.addAction(android.R.drawable.ic_menu_close_clear_cancel,
          mContext.getString(R.string.notification_action_cancel),
          PendingIntent.getBroadcast(mContext, 0, cancelIntent, 0))
      builder.setCategory(NotificationCompat.CATEGORY_PROGRESS)
    } else if (type == NotificationStatus.COMPLETE) {
      val record = cluster.firstOrNull()
      if (record != null && id != -1L) {
        // TODO: add redo for failed uploads, and set Visibility accordingly
        val uri = ContentUris.withAppendedId(UploadContract.UPLOAD_CONTENT_URI, id)
        builder.setAutoCancel(true)
        val action =
            if (record.Status.isFailed()) {
              UploadContract.NotificationAction.List
            } else {
              UploadContract.NotificationAction.Open
            }

        val intent = Intent(action.actionString, uri, mContext, UploadReceiver::class.java)
        intent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND)
        intent.putExtra(UploadManager.EXTRA_NOTIFICATION_CLICK_UPLOAD_IDS, uploadIds)
        builder.setContentIntent(PendingIntent.getBroadcast(mContext,
            0, intent, PendingIntent.FLAG_UPDATE_CURRENT))

        val hideIntent = Intent(UploadContract.NotificationAction.Hide.actionString,
            uri, mContext, UploadReceiver::class.java)
        hideIntent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND)
        builder.setDeleteIntent(PendingIntent.getBroadcast(mContext, 0, hideIntent, 0))
      } else {
        Log.d(TAG, "The cluster of $tag is empty.")
      }
    }

    val (percent, remainingText, percentText) = buildActiveInfo(cluster)
    if (type == NotificationStatus.ACTIVE) {
      // set progress bar if it's active upload
      builder.setProgress(100, Math.max(0, percent), percent < 0)
    }

    val notification: Notification
    if (cluster.size == 1) {
      val record = cluster.first()
      builder.setContentTitle(getUploadTitle(res, record))
      when (type) {
        NotificationStatus.ACTIVE -> {
          if (!record.NotificationDescription.isNullOrEmpty()) {
            builder.setContentText(record.NotificationDescription)
          } else {
            builder.setContentText(remainingText)
          }
          builder.setContentInfo(percentText)
        }
        NotificationStatus.WAITING -> {
          builder.setContentText(
              resources.getString(R.string.notification_waiting_for_suitable_network)
          )
        }
        NotificationStatus.COMPLETE -> {
          when {
            record.Status.isSuccess() -> {
              builder.setContentText(resources.getString(R.string.notification_upload_successfully))
            }
            record.Status.isDeletedOrCanceled() -> {
              builder.setContentText(resources.getString(R.string.notification_upload_cancel))
            }
            record.Status.isFailed() -> {
              builder.setContentText(resources.getString(R.string.notification_upload_unsuccessfully))
            }
          }
        }
      }
      notification = builder.build()
    } else {
      val inboxStyle = NotificationCompat.InboxStyle(builder)
      cluster.forEach { inboxStyle.addLine(getUploadTitle(res, it)) }

      if (type == NotificationStatus.WAITING) {
        builder.setContentTitle(
            res.getQuantityString(R.plurals.notif_summary_waiting, cluster.size, cluster.size))
        builder.setContentText(
            res.getString(R.string.notification_waiting_for_suitable_network))
        inboxStyle.setSummaryText(
            res.getString(R.string.notification_waiting_for_suitable_network))
      }
      notification = inboxStyle.build()
    }

    mNotifManager.notify(tag.toString(), 0, notification)
  }

  // <percent, remainingText, percentText>
  private fun buildActiveInfo(cluster: Iterable<UploadRecord>): Triple<Int, String?, String?> {
    var remainingText: String? = null
    var percentText: String? = null
    var percent = -1

    var current: Long = 0
    var total: Long = 0
    var speed: Long = 0
    synchronized(mUploadSpeed) {
      val knownSizeCluster = cluster.filter { it.TotalBytes != -1L }
      current = knownSizeCluster.sumBy { it.CurrentBytes.toInt() }.toLong()
      total = knownSizeCluster.sumBy { it.TotalBytes.toInt() }.toLong()
      speed = knownSizeCluster.sumBy { mUploadSpeed[it._ID]?.first?.toInt() ?: 0 }.toLong()
    }

    if (total > 0) {
      percentText = NumberFormat.getPercentInstance().format(current.toDouble() / total)

      if (speed > 0) {
        val remainingMillis = (total - current) * 1000 / speed

        remainingText = resources.getString(R.string.upload_remaining,
            remainingMillis.formatDuration(resources))
      }

      percent = (current * 100 / total).toInt()
    }

    return Triple(percent, remainingText, percentText)
  }

  companion object {

    private object InstanceHolder : SingletonHolder<UploadNotifier, Context>(::UploadNotifier)

    internal val getInstance = InstanceHolder::getInstance

    private const val CHANNEL_ACTIVE = "active"
    private const val CHANNEL_WAITING = "waiting"
    private const val CHANNEL_COMPLETE = "complete"

    private val TAG = LogUtils.makeTag<UploadNotifier>()

    private fun getUploadTitle(res: Resources, info: UploadRecord): CharSequence {
      val title = info.NotificationTitle
      return if (title.isNullOrEmpty()) {
        // TODO: fix for multiple files
        info.Parts.firstOrNull { it.fileInfo != null }?.fileInfo?.fileName
            ?: res.getString(R.string.upload_unknown_upload_title)
      } else {
        title
      }
    }

    private fun getUploadIds(infos: Iterable<UploadRecord>): LongArray {
      return infos.map { it._ID }.toLongArray()
    }
  }

  internal enum class NotificationStatus {
    WAITING,
    ACTIVE,
    COMPLETE
  }
}
